app/src/main/java/org/dhis2/utils/category/CategoryDialogPresenter.kt

```diff
@@ -0,0 +1,192 @@
+import org.dhis2.data.schedulers.SchedulerProvider
+import org.hisp.dhis.android.core.D2
+import org.hisp.dhis.android.core.arch.repositories.scope.RepositoryScope
+import org.hisp.dhis.android.core.category.CategoryOption
+import org.hisp.dhis.android.core.category.CategoryOptionCombo
+import timber.log.Timber
+
+class CategoryDialogPresenter(
+    val view: CategoryDialogView,
+    val d2: D2,
+    val type: CategoryDialog.Type,
+    val uid: String,
+    private val withAccessControl: Boolean,
+    private val date: Date?,
+    private val catOptMapper: CategoryOptionCategoryDialogItemMapper,
+    private val catOptCombMapper: CategoryOptionComboCategoryDialogItemMapper,
+    val schedulerProvider: SchedulerProvider
+) {
+
+    var disposable: CompositeDisposable = CompositeDisposable()
+
+    fun init() {
+        when (type) {
+            CategoryDialog.Type.CATEGORY_OPTIONS -> getOptions()
+            CategoryDialog.Type.CATEGORY_OPTION_COMBO -> getCategoryOptionCombos()
+        }
+    }
+
+    private fun getCategoryOptionCombos() {
+        disposable.add(
+            d2.categoryModule().categoryCombos().uid(uid).get()
+                .subscribeOn(schedulerProvider.io())
+                .observeOn(schedulerProvider.ui())
+                .subscribe(
+                    { view.setTitle(it.displayName() ?: "-") },
+                    { Timber.e(it) }
+                )
+        )
+
+        disposable.add(
+            view.searchSource()
+                .debounce(500, TimeUnit.MILLISECONDS, schedulerProvider.io())
+                .map { it.toString() }
+                .map { textToSearch ->
+                    mapCatOptComboToLivePageList(textToSearch)
+                }
+                .subscribeOn(schedulerProvider.io())
+                .observeOn(schedulerProvider.ui())
+                .subscribe(
+                    { view.setLiveData(it) },
+                    { Timber.e(it) }
+                )
+        )
+    }
+
+    private fun createDataSource(
+        dataSource: DataSource<CategoryOptionCombo, CategoryDialogItem>
+    ): LiveData<PagedList<CategoryDialogItem>> {
+        return LivePagedListBuilder(
+            object : DataSource.Factory<CategoryOptionCombo, CategoryDialogItem>() {
+                override fun create(): DataSource<CategoryOptionCombo, CategoryDialogItem> {
+                    return dataSource
+                }
+            },
+            20
+        ).build()
+    }
+
+    private fun getOptions() {
+        disposable.add(
+            d2.categoryModule().categories().uid(uid).get()
+                .subscribeOn(schedulerProvider.io())
+                .observeOn(schedulerProvider.ui())
+                .subscribe(
+                    { view.setTitle(it.displayName() ?: "-") },
+                    { Timber.e(it) }
+                )
+        )
+        disposable.add(
+            view.searchSource()
+                .debounce(500, TimeUnit.MILLISECONDS, schedulerProvider.io())
```

Comment:
why debounce?

Reply:
Because i just want to apply the result when the user stops writting